### PR TITLE
feat: add more http-server initialization options

### PR DESCRIPTION
**Pull request by LautaroPetaccio**
## what

adds more options to `createServerComponent` in `@dcl/http-server`.

- **`maxBodySize`** (bytes): rejects oversized request bodies with `413 payload too large`. enforced up-front via `content-length` (before the body is read) and, for bodies that omit or under-declare their length (e.g. chunked transfer-encoding), while streaming. unset means no limit.
- **`requestTimeout`** → `server.requestTimeout`
- **`maxHeadersCount`** → `server.maxHeadersCount`
- **`maxRequestsPerSocket`** → `server.maxRequestsPerSocket`

when `maxBodySize` is unset the request body path is unchanged.

## bonus fix

fixes a latent bug in `getServer` where passing `https` options was silently overwritten by the plain-http fallback (`if (https) … if (http) … else …`), so `https` servers were never actually created. now an `else if` chain.

## tests

- unit tests for the `getServer` tunables + https fix, `exceedsContentLength`, and the streaming `createBodySizeLimiter`.
- e2e: within-limit echoes `200`, over-limit `content-length` returns `413`/`payload too large`, bodyless `GET` unaffected.

full suite passes (355 passed, 4 skipped).

## changeset

`minor` bump for `@dcl/http-server`.

### CODE REVIEW

File: components/http-server/src/server.ts [hunk 1]
```diff
@@ -111,7 +112,19 @@ export async function createServerComponent<Context extends object>(
...
+    // Bodies that omit or under-declare `Content-Length` are still capped while streaming by the
+    // limiter in `getRequestFromNodeMessage`.
+    if (maxBodySize !== undefined && exceedsContentLength(req.headers['content-length'], maxBodySize)) {
+      res.statusCode = 413
+      res.setHeader('content-type', 'text/plain; charset=utf-8')
+      // Close the connection: the declared body is never read, so the socket can't be reused.
+      res.setHeader('connection', 'close')
+      res.end('Payload Too Large')
+      return
+    }
```
[the remark this answers is not in the document]
Comment (LautaroPetaccio) [reply components/http-server/src/server.ts:159]: Thanks — I dug into this and I believe it's a **false positive**. The code already sets `Connection: close` on the early 413, which makes Node flip `res.shouldKeepAlive = false` and tear the socket down server-side after the response finishes, regardless of the unconsumed request body. `Connection: close` is *not* only a client-side hint here.

I verified with a raw-socket slowloris (huge `Content-Length`, no body sent), comparing three variants of this exact branch:

```
NO Connection:close, NO destroy   => STILL-OPEN @5000ms   (hangs; requestTimeout is already cleared once a response is sent)
Connection:close,    NO destroy   => closed @2ms          (current code)
NO Connection:close, WITH destroy => closed @1ms
```

So the socket is freed in ~2ms as-is. I deliberately did **not** add the suggested synchronous `res.end('Payload Too Large'); req.destroy()` — destroying the socket synchronously after `end()` can truncate the response before it flushes (the same reason the streaming limiter doesn't destroy its source). Happy to add a `req.destroy()` inside the `res.end(..)` finish callback as defense-in-depth if you'd prefer, but it's redundant given `Connection: close` already handles teardown.

**Comment by LautaroPetaccio** [comment]
Pushed `821bff7` addressing the review.

**P1 (early 413 doesn't free the socket)** — false positive, see [inline reply](https://github.com/decentraland/core-components/pull/128#discussion_r3469402144): `Connection: close` already tears the socket down server-side in ~2ms (verified with a raw-socket slowloris). No change.

**Addressed:**
- **P2 (validate `maxBodySize`)** — `createServerComponent` now throws on `0`, negative or fractional values (omit for "no limit").
- **P2 (CORS on early 413)** — extracted `getActualResponseCorsHeaders` (shared with the CORS middleware); the up-front 413 now carries actual-response CORS headers when `cors` is configured. Covered for wildcard and reflected-origin.
- **P2 (doc the `>` boundary)** — documented the inclusive limit in the type/JSDoc.
- **P2 (test coverage)** — added `exceedsContentLength` edge cases (`-1`, `Infinity`, `NaN`) and an end-to-end chunked (no `Content-Length`) over-limit test that confirms a 413.

Full suite: 366 passed, 4 skipped.

**Two known limitations I'm leaving as-is (callouts, not fixes):**
1. The early `Content-Length` rejection responds before the middleware chain, so it won't appear in request-logger/metrics middlewares — 413s rejected up-front are undercounted in those.
2. The streaming-limiter 413 (chunked / lying `Content-Length`) goes through the normal response path without `Connection: close`, so Node may read-and-discard the rest of the body to reuse the socket. Memory stays bounded (~16KB via the limiter's backpressure), but a deliberately-unbounded chunked stream could keep the server reading. The up-front check covers all honest clients; happy to harden this path (force `Connection: close` on body-limit 413s) if we think it's worth it.

**Comment by LautaroPetaccio** [comment]
Pushed `c103fc2` — followed up on whether `maxBodySize` applies to multipart requests.

**Yes, with a caveat I found and fixed:**

- **Normal multipart uploads** (browser/`fetch` `FormData`, which set `Content-Length`) are rejected up-front with a clean `413` — the `Content-Length` check is content-type agnostic. ✓
- **Chunked multipart** (no `Content-Length`) parsed via the busboy wrapper hit a gap: the limiter fires its `413` on the body stream, but `Readable.fromWeb(ctx.request.body).pipe(parser)` doesn't forward *source* errors to `parser`, so it surfaced as an **unhandled error** (→ `400` to the client, and a crash risk). This was actually a pre-existing gap in that pipe pattern — a plain client-abort mid-upload would go unhandled the same way; `maxBodySize` just makes it attacker-triggerable.

**Fix:**
- Forward body-stream errors to the parser in the multipart wrapper (`body.on('error', (err) => parser.destroy(err))`), so the `413` propagates cleanly through the error middleware.
- Documented in the README that consumers **piping** `ctx.request.body` must handle stream errors (abort *or* the `maxBodySize` `413`), with the correct snippet. Reading via `await ctx.request.json()/.text()` or `for await` already propagates automatically.
- Added multipart-over-limit tests for both the `Content-Length` and chunked paths.

Full suite: 368 passed, 4 skipped.

**Comment by LautaroPetaccio** [comment]
Pushed `7301dbf` — fixed the metrics mislabelling I found while answering the "do we have metrics for these cases?" question.

**Problem:** the metrics middleware (`instrumentHttpServerWithPromClientRegistry`) is registered *inside* `coerceErrorsMiddleware`, so a **thrown** error reached it as an exception (`res` undefined) and was counted as `code="200"` before `coerceErrorsMiddleware` converted it to a response. This affected the `maxBodySize` streaming 413 **and every other thrown error** (a `throw createError(404)` → counted 200; a `throw new Error()` → 200 instead of 500). Returned responses were already labelled correctly.

**Fix:** the middleware now derives the status from the thrown error (`err.status ?? err.statusCode ?? 500`) in a `catch`, so `http_requests_total`, `http_request_duration_seconds` and `http_request_size_bytes` all carry the real code. Added tests for returned-status, thrown-http-error (413), and thrown-plain-error (500).

**Heads-up for dashboards:** error traffic that previously showed up under `code="200"` will now move to its real status code. That's the correct behaviour, but worth a glance at any alerts keyed on the old counts.

**Still not captured (by design):** the up-front `Content-Length` rejection responds before the middleware chain, so it isn't in app metrics — it's a transport-layer reject (observable at the LB). Documented in the README; left as-is per the earlier discussion.

Full suite: 371 passed, 4 skipped.

**Comment by LautaroPetaccio** [comment]
Pushed `f087056` — the up-front `Content-Length` rejection now logs a warning.

It responds before the middleware chain (so it can't be captured by the metrics middleware) and was previously silent. The component already has a logger, so this is zero new coupling:

```
logger.warn('Rejected request: body exceeds maxBodySize', { method, contentLength, maxBodySize })
```

This makes the rejection observable for log-based alerting without pulling a metrics dependency into the core server. Added a test asserting the warning is emitted with the method and configured limit.

Full suite: 372 passed, 4 skipped.

**Comment by LautaroPetaccio** [comment]
Pushed `aac19e9` — extended body-size limiting to `@dcl/schema-validator-component`.

`createSchemaValidatorComponent({ maxBodySize })` (bytes): the validation middleware now rejects a request that declares a larger `Content-Length` with `413 Payload Too Large` **before** parsing its body. The value is validated as a positive integer at construction, mirroring the http-server option.

It's intentionally a **declared-size (Content-Length) guard**, not a streaming reader — the component is a thin ajv wrapper, and the robust streaming/chunked cap is the transport layer's job. The README documents pairing it with the `@dcl/http-server` `maxBodySize` for bodies that omit/under-declare their length. Added tests (over-limit → 413 without parsing, within-limit → proceeds, no-Content-Length → proceeds, and construction validation) and a separate changeset (`@dcl/schema-validator-component` minor).

Heads-up: this PR now spans two packages (http-server + schema-validator). Both are the same "body-size limiting" theme, but happy to split the schema-validator out into its own PR if you'd prefer to keep #128 http-server-only — let me know.

File: components/http-server/src/metrics.ts [hunk 2]
```diff
@@ -131,6 +131,13 @@ export async function instrumentHttpServerWithPromClientRegistry<K extends strin
 
     try {
       return (res = await next())
+    } catch (error: any) {
+      // A thrown error is turned into a response by `coerceErrorsMiddleware`, which runs *outside*
+      // this middleware, so the final response object never reaches us here. Derive the status the
+      // same way it does, so the metric is labelled with the real code (e.g. a thrown `413` or
+      // `404`, or `500` for an unmapped error) instead of defaulting to `200`.
+      labels.code = error?.status || error?.statusCode || 500
```
[the remark this answers is not in the document]
Comment (LautaroPetaccio) [reply components/http-server/src/metrics.ts]: Good catch — fixed in `0bcab65`. Changed it to `labels.code = Number(error?.status || error?.statusCode) || 500`, so a non-numeric `.status`/`.statusCode` coerces to `NaN` and falls back to `500` instead of leaking a string into the Prometheus label. Added a regression test (`throw { status: 'totally-not-a-number' }` → labelled `500`).

File: components/schema-validator/src/component.ts [hunk 3]
```diff
@@ -63,6 +68,20 @@ export function createSchemaValidatorComponent<T extends object>(options?: {
         }
       }
 
+      if (maxBodySize !== undefined) {
+        const contentLengthHeader = context.request.headers.get('Content-Length')
```
[the remark this answers is not in the document]
Comment (LautaroPetaccio) [reply components/schema-validator/src/component.ts]: Agreed — fixed in `0bcab65`. I made the limitation prominent in the `maxBodySize` JSDoc (a ⚠️ block stating it is **NOT a streaming guard**, that `.json()` still buffers the full body, and that the `@dcl/http-server` `maxBodySize` is what actually caps bytes off the socket) plus a note at the check site. I skipped the console-warning idea: the schema-validator has no reference to the transport layer, so it can't reliably detect whether a transport-level cap exists — a warn would either be noisy on every construction or produce false positives.

**Comment by LautaroPetaccio** [comment]
Pushed `d08bd36` — added a reusable per-route body-size-limit middleware to `@dcl/http-server`, as the cleaner home for a true streaming cap (instead of reimplementing stream-capping inside the ajv wrapper).

```ts
import { createBodySizeLimitMiddleware } from '@dcl/http-server'

router.post('/v1/notes', createBodySizeLimitMiddleware(4096), notesHandler)
```

**How it enforces the limit (dual check, same as the server-wide option):**
- A request declaring a larger `Content-Length` → `413 Payload Too Large` up front, before `next()`.
- A body that omits/under-declares its length (chunked) → routed through a web `TransformStream` that errors with a `413` once the limit is crossed; the middleware replaces `ctx.request` so any downstream read flows through the cap. The error propagates as a `413` via the error middleware.

**Why a middleware (vs. baking streaming into the schema-validator):** while exploring the validator route I hit a real web-streams gotcha — `reader.cancel()` on a `clone()`d body **hangs** (tee waits for both branches) and tee buffers the full body for the unread branch, so cloning can't bound memory. Capping at the request level here avoids that, keeps the validator a thin ajv wrapper, and is reusable per-route. The schema-validator JSDoc now points here for the streaming cap.

Notes:
- Composes with the server-wide `maxBodySize` (global cap applies first). Size must be a positive integer or the factory throws.
- Tests follow the testing standards; full http-server suite 391 passed / 4 skipped. Changeset added (`@dcl/http-server` minor).

**Comment by LautaroPetaccio** [comment]
Pushed `c3d1cfa` — wired the schema-validator → body-size-limiter interaction.

**Problem:** the validator wraps `context.request.clone().json()` in a try/catch that mapped *any* error to `400`. So when the `@dcl/http-server` body-size limiter (server-wide `maxBodySize` or `createBodySizeLimitMiddleware`) capped a chunked/under-declared body, the `413` it raised on the body read surfaced as a `400`.

**Fix:** the catch now duck-types an HTTP `status`/`statusCode` off the thrown error (same convention as `coerceErrorsMiddleware`, no new dependency on the http layer) and surfaces it. A `413` from an upstream limiter stays a `413`; a genuine JSON parse error (no status) still returns `400`.

Verified the crux first: a `413` raised by the limiter's web `TransformStream` survives `pipeThrough` → rebuilt `Request` → the validator's `.clone().json()` with `status: 413` intact. Added unit tests for both the `.status` and `.statusCode` branches (and the existing no-status → `400` test still passes). 24 passed.

I did not add a cross-package integration test (it'd mean a `@dcl/http-server` dev-dependency + build-order coupling in the schema-validator package); both sides of the contract are pinned by committed tests — the limiter spec asserts the body read rejects with `status: 413`, and the validator spec asserts a statused error is surfaced. Changeset added (`@dcl/schema-validator-component` patch).

**Comment by LautaroPetaccio** [comment]
Pushed `aeee1be` — removed the `maxBodySize` option from `@dcl/schema-validator-component`, since body-size limiting now lives at the transport layer (`@dcl/http-server`'s server-wide `maxBodySize` or `createBodySizeLimitMiddleware`).

Removed:
- the `maxBodySize` option + its construction validation + the `Content-Length` pre-parse check
- the `SchemaValidatorOptions` type (reverted to the inline `{ ensureJsonContentType }` options)
- the related tests and the `schema-validator-max-body-size` changeset

**Kept** (still valuable, arguably more so now): the validator honors an upstream body-read error that carries an HTTP `status`/`statusCode` — so a `413` raised by the http-server limiter surfaces as a `413` instead of being masked as a `400`. Its tests and changeset stay.

Net effect: the schema-validator's only change in this PR is now that `httpErrorStatus` handling. README updated to point at the http-server limiters for bounding body size. Suite: 16 passed.

**Comment by LautaroPetaccio** [comment]
Pushed `0e97532` — security review of the body-size changes turned up a real **socket-exhaustion DoS** that I've now fixed.

**Finding:** an oversized-body `413` left the socket open. Two exploitable shapes, both verified with raw sockets:
- *Slowloris*: declare a large `Content-Length` and send nothing → the middleware's early `413` didn't close the socket → **STILL-OPEN @4s** (the request never completes and `requestTimeout` is cleared once a response is sent).
- *Infinite chunked*: stream a chunked body past the limit → the cap errored but the `413` went out (or didn't) without `Connection: close`, and the client kept streaming → **STILL-OPEN @4s**. This affected **both** the per-route middleware **and** the server-wide `maxBodySize`.

**Fix:** every body-size `413` now sets `Connection: close`.
- Middleware: the up-front `413` sets it; the streaming path is caught in the middleware and returns a `413`+close (instead of letting the error propagate uncosed).
- Server-wide: an `onExceeded` signal is threaded through the Node-stream limiter so `asyncHandle` adds `Connection: close` to the streaming `413` too.

After the fix both shapes get a `413` and the socket closes in **<100ms**. Added regression tests (unit assertions on the `connection: close` header for both paths, plus the raw-socket e2e). Suites: http-server **402**, schema-validator **16**.

**Also reviewed, no action needed:**
- The earlier metrics fix (`Number(...) || 500`) already prevents a high-cardinality Prometheus label from a non-numeric thrown status — a stability/DoS concern.
- The exposed `requestTimeout` / `maxHeadersCount` / `maxRequestsPerSocket` tunables are themselves slowloris/DoS knobs.
- Perf: the limiter passes chunks through by reference (no data copy); the early `Content-Length` check rejects before building the `Request`; the throwaway CORS `Request` is only built on the (rare) rejection path. Nothing hot-path-regressing.
